### ISSUE TITLE
Implement TestRunParameters inside .testsettings for runtime parameters

### ISSUE THREAD

**Issue by FermJacob** [nunit/nunit3-vs-adapter#145]
[Reference #49 ](https://github.com/nunit/nunit3-vs-adapter/issues/49#issuecomment-201333589)
[Blocked by #720](https://github.com/nunit/nunit/issues/720)

Scenario:
I am using this with Selenium tests to override which browser to use and the URL of the web application at runtime for multiple environments.  I am looking to switch to NUnit because VSTest runs tests in parallel at the assembly not the class/test(I know test hasn't been implemented) like NUnit to speed up execution time.  

It is desired to implement TestRunParameters inside the .runsettings file into the VS Adapter for runtime changes.  VSTest with Visual Studio (Update 1?) and VSTS allow for these parameters to be used.  

[MSDN](https://msdn.microsoft.com/en-us/library/jj635153.aspx).  They state inside the ForcedLegacyMode "We recommend that you consider refactoring your tests to allow you to use the newer adapter."  Here is the original blog post about the TestRunParameters: [Blog](https://blogs.msdn.microsoft.com/visualstudioalm/2015/09/04/supplying-run-time-parameters-to-tests/) .  It gives a good example on how to implement it and override the parameter with VSTest at runtime with the VSTS builds.

Example XML for .runsettings file:
```xml
<TestRunParameters>
	<Parameter name="webAppUrl" value="https://google.com" />
	<Parameter name="Username" value="username" />
	<Parameter name="Password" value="password" />
</TestRunParameters>
```

Example code for MS Unit Test Framework:
```csharp
public static string GetString(TestContext context, string property)
{
        if (context.Properties[property] == null)
        {
            throw new ArgumentException("Property does not exist or does not have a value");
        }
        else
        {
            return context.Properties[property].ToString();
        }
}
```

**Comment by CharliePoole** [comment]
Thanks for filing this - it will help to track things.

**Comment by paululvinius** [comment]
+1, we definitely need to be able to read custom properties. Surprised this is lacking in NUnit, it's a show stopper for us since we also use configuration through automation tools to set app urls (UI tests and API tests). The already existing TestContext class should have a TestContext.Properties[] that gets filled from command line params or xml config. Once this feature is provided, VSTS and other tools can use this to fully integrate with NUnit.

**Comment by CharliePoole** [comment]
Yes, adding a Properties element to the TestContext makes sense. Do you feel it needs to be an array? I would lean towards an IDictionary.

One decision point is whether it should  be a static member of TestContext or an instance member. In the latter case, you would access TestContext.CurrentContext.Properties. We should discuss the pros and cons on the nunit issue.

**Comment by paululvinius** [comment]
It should probably be an IDictionary<string, string>. (MSTest uses a non-generic IDictionary but that’s probably for legacy reasons).

Since these values won't change between tests I see no reason to use CurrentContext. The dictionary should be global scope (e.g. through a singelton instance) and probably be read only (once a test run has started it shouldn't  be used for storing state between tests.)
An instance member like CurrentContext.Properties might serve a different purpose though. It's likely to have different behaviors and constraints, so for this we're probably looking at another feature that should be discussed in a separate issue.

**Comment by CharliePoole** [comment]
If it's global, I think it has to be read-only. Otherwise you have people saving things into it that affect tests that happen to run later - possibly even completely unrelated tests.

If it's an instance member, you could let the one time setup add values, then copy the dictionary for each child. That would give you one-way comunication down the tree of tests without anything at a lower level being able to mess up the higher level.

However, this is all for the design within NUnit, not for the adapter.

**Comment by CharliePoole** [comment]
This continues to be blocked. The feature may be included in NUnit 3.4, at which point this issue will be available for working.

**Comment by lucasBertola** [comment]
+1
